src/psyclone/domain/nemo/transformations/nemo_kernel_trans.py

```diff
@@ -0,0 +1,127 @@
+    def name(self):
+        '''
+        :returns: the name of the transformation.
+        :rtype: str
+
+        '''
+        return type(self).__name__
+
+    def validate(self, node, options=None):
+        '''
+        Check that the supplied node is a valid target for this transformation.
+
+        :param node: the target of the transformation.
+        :type node: :py:class:`psyclone.psyir.nodes.Node`
+        :param options: a dictionary with options for \
+            transformations. No options are used in this \
+            transformation. This is an optional argument that defaults \
+            to None.
+        :type options: dict of string:values or None
+
+        :raises TransformationError: if the supplied node is not a Schedule, \
+            is not within a loop or cannot be represented as a Kernel.
+
+        '''
+        super(NemoKernelTrans, self).validate(node)
+
+        if not isinstance(node, Schedule):
+            raise TransformationError(
+                "Error in NemoKernelTrans transformation. The supplied node "
+                "should be a PSyIR Schedule but found '{0}'".format(
+                    type(node).__name__))
+
+        # A Kernel must be within a Loop
+        if not isinstance(node.parent, Loop):
+            raise TransformationError(
+                "Error in NemoKernelTrans transformation. The supplied "
+                "Schedule must be within a Loop.")
+
+        # Check for array assignments
+        nodes = [assign for assign in node.walk(Assignment)
+                 if assign.is_array_range]
+        if nodes:
+            fwriter = FortranWriter()
+            raise TransformationError(
+                "A NEMO Kernel cannot contain array assignments but found: "
+                "{0}".format([fwriter(node).rstrip("\n") for node in nodes]))
+
+        # A kernel cannot contain loops, calls or unrecognised code (including
+        # IO operations. So if there is any node in the result of
+        # the walk, this node cannot be represented as a NEMO kernel.
+        nodes = node.walk((CodeBlock, Loop, Call))
+        if nodes:
+            raise TransformationError(
+                "Error in NemoKernelTrans transformation. A NEMO Kernel cannot"
+                " contain nodes of type: {0}".format(
+                    [type(node).__name__ for node in nodes]))
+
+    def apply(self, sched, options=None):
+        '''
+        Takes a generic PSyIR Schedule and replaces it with a NEMO Kernel.
+
+        :param sched: the Schedule node to be transformed.
+        :type sched: :py:class:`psyclone.psyir.nodes.Schedule`
+        :param options: a dictionary with options for \
+            transformations. No options are used in this \
+            transformation. This is an optional argument that defaults \
+            to None.
+        :type options: dict of string:values or None
+
+        '''
+        self.validate(sched)
+
+        nemokern = NemoKern(sched.pop_all_children(),
```

Comment:
If I am not wrong the second parameter is never used for NemoKern (is always `None`), can it be deleted?

Reply:
That was a mistake - 'options' should have been passed into `validate()` so that it can potentially be picked up by a base class.

Reply:
Sorry for not being clear here, I meant the second attribute on the NemoKern constructor. Which is declared at nemo.py:357 Using grep I see this is always None, but maybe I am missing some use? I guess the _ast if assigned by the Fortran frontend instead than the constructor?

Also looking then at the NemoKern class it seems the "pylint disable" is no longer needed?

Reply:
Ah, no, you were clear - I just misread it! You are right, I'm able to remove the fparser2 AST argument now and have gone through the code and tidied up. Felt good :-)